apis/traffic-access/v1alpha4/traffic-access.md

```diff
@@ -0,0 +1,352 @@
+metadata:
+  name: api-service-api
+  namespace: default
+spec:
+  destination:
+    kind: IdentityBinding
+    group: access.smi-spec.io
+    name: api-service
+    namespace: default
+  rules:
+  - kind: TCPRoute
+    name: api-service-port
+  - kind: HTTPRouteGroup
+    name: api-service-routes
+    matches:
+    - api
+  sources:
+  - kind: IdentityBinding
+    group: access.smi-spec.io
+    name: website-service
+    namespace: default
+  - kind: IdentityBinding
+    group: access.smi-spec.io
+    name: payments-service
+    namespace: default
+```
+
+The previous example would allow the following HTTP traffic:
+
+| source            | destination   | path     | method |
+| ----------------- | ------------- | -------- | ------ |
+| website-service   | api-service   | /api     | *      |
+| payments-service  | api-service   | /api     | *      |
+| prometheus        | api-service   | /metrics | GET    |
+
+## Example implementation for L4
+
+The following implementation shows how to define TrafficTargets for
+allowing TCP and UDP traffic to specific ports.
+
+```yaml
+kind: TCPRoute
+metadata:
+  name: tcp-ports
+spec:
+  matches:
+    ports:
+    - 8301
+    - 8302
+    - 8300
+---
+kind: UDPRoute
+metadata:
+  name: udp-ports
+spec:
+  matches:
+    ports:
+    - 8301
+    - 8302
+---
+kind: TrafficTarget
+metadata:
+  name: protocal-specific
+spec:
+  destination:
+    kind: IdentityBinding
+    group: access.smi-spec.io
+    name: server
+    namespace: default
+  rules:
+  - kind: TCPRoute
+    name: tcp-ports
+  - kind: UDPRoute
+    name: udp-ports
+  sources:
+  - kind: IdentityBinding
+    group: access.smi-spec.io
+    name: client
+    namespace: default
+```
+
+Note that the above configuration will allow TCP and UDP traffic to
+both `8301` and `8302` ports, but will block UDP traffic to `8300`.
+
+## Tradeoffs
+
+- Additive policy - policy that denies instead of only allows is valuable
+  sometimes. Unfortunately, it makes it extremely difficult to reason about what
+  is allowed or denied in a configuration.
+
+- Resources vs selectors - it would be possible to reference concrete resources
+  such as a deployment instead of selecting across pods.
+
+- This access control is on the destination (server) side, implicitly
+
+- Currently the specification does not have provision for the definition of
+  higher level elements such as a service. It is probable that this specification
+  will change once these elements are defined.
+
+## Out of scope
+
+- Egress policy - TrafficTarget does *not* allow for the possibility of egress
+  access control as it selects specific pods and not hostnames. Another object
+  will need to be created to manage this use case.
+
+- Ingress policy - assuming clients present the correct identity, this *should*
+  work for some kind of ingress. Unfortunately, it does not cover many of the
+  common use cases (filtering by hostname) and will need to be expanded to cover
+  this use case.
```

Comment:
This might be a relevant case to consider proposing `TrafficTarget` at some point in the future (maybe with a different name to better reflect its authorization purpose?) as an extension to the [Kubernetes Gateway API](https://gateway-api.sigs.k8s.io/), which I believe intends to support some forms of authentication, but AFAIK doesn't yet have any way of representing those (JWT identity is probably something that would be desirable to add at that point).